Enhance UTF-8 support for filename handling in downloads

arash77 opened the pull request:
Fixes step 3 of #18584
Improve filename handling in data downloads and zipstream headers to support UTF-8 encoding, ensuring proper display of filenames with special characters.

## How to test the changes?
(Select all options that apply)
- [x] I've included appropriate [automated tests](https://docs.galaxyproject.org/en/latest/dev/writing_tests.html).
- [ ] This is a refactoring of components with existing test coverage.
- [ ] Instructions for manual testing are as follows:
  1. [add testing steps and prerequisites here if you didn't write automated tests covering all your changes]

## License
- [x] I agree to license these and all my past contributions to the core galaxy codebase under the [MIT license](https://opensource.org/licenses/MIT).

arash77 committed:
Enhance filename handling in data download and zipstream headers to support UTF-8 encoding

arash77 committed:
Add tests for datasets and collections download with non english characters

davelopez commented on lib/galaxy/datatypes/data.py:
```diff
@@ -430,14 +431,16 @@ def _serve_raw(
         )
-        headers["Content-Disposition"] = f'attachment; filename="{filename}"'
+        headers["Content-Disposition"] = (
+            f"attachment; filename=\"{filename}\"; filename*=UTF-8''{utf8_encoded_filename}"
```
Can we move this to a utility function to reduce duplication?
Something like `def to_content_disposition(filename: str)` or something similar. The utf8 encoding should be inside too, so `_download_filename` just returns `filename`.

arash77 replied:
Thanks for the suggestion!
Should I remove the `escape` from `name` in `_download_filename`?
Is it even necessary to escape the name anymore? Can we just use the quote function to escape?

davelopez replied:
If it is not relevant anymore I guess so? it seems to be used in `hdca_name` so check carefully if you can really drop it :+1:

arash77 replied:
I dropped the `escape` function and also removed it from `hdca_name` too, as it didn’t seem necessary since the quote function handles it effectively.

mvdbeek replied:
The escape function limits input to 150 characters and excludes characters that aren't file system safe.

arash77 replied:
We can limit the characters and I think that the `quote` function already excludes these kinds of characters.

mvdbeek replied:
No it does not, it just encodes them and the user-agent then knows how to turn them back into the original form.

mvdbeek replied:
See https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Content-Disposition#filename, which unfortunately doesn't go into detail on `restricted characters`

arash77 replied:
So we have to keep this but I will move this part into the new function.

arash77 committed:
Refactor filename handling in data download to use a dedicated toContentDisposition function for improved UTF-8 support

arash77 committed:
Sanitize filenames in toContentDisposition function to ensure valid characters are used

mvdbeek commented on lib/galaxy/util/__init__.py:
```diff
@@ -2006,3 +2007,9 @@ def lowercase_alphanum_to_hex(lowercase_alphanum: str) -> str:
+
+
+def toContentDisposition(filename: str) -> str:
+    sanitized_filename = "".join(c in FILENAME_VALID_CHARS and c or "_" for c in filename)[0:150]
```
You will also need to apply the file length limit and exclude certain characters in`utf8_encoded_filename` as well.

davelopez replied:
and the name should be `to_content_disposition` or similar, sorry my brain was in frontend mode when I suggested the change :smile:

arash77 replied:
ok, also the limit is applied to 150 characters shouldn't it be 255?

arash77 replied:
And by `exclude certain characters` you mean which characters?
Should this be excluded? `/\?%*:|"<>`

mvdbeek replied:
https://github.com/galaxyproject/galaxy/pull/19161#discussion_r1850153897

arash77 replied:
I couldn't find any specific limitation on characters from the link but I found out that for a typical filename, the characters limitation is 256 characters and `/\?%*:|"<>` are not allowed.

mvdbeek replied:
Sounds good. I bet most browsers are putting proper restrictions in place anyway.

arash77 committed:
Rename toContentDisposition function to to_content_disposition
apply 255 characters limit to filename
exclude `/\?%*:|"<>` characters from encoded filename

arash77 committed:
exclude ext from character_limit
